### PR TITLE
Improve search code

### PR DESCRIPTION
**Pull request by Nikerabbit**
* Always use searchAPI if provided
* Clean up code flow to be easier to follow

https://phabricator.wikimedia.org/T73891

### CODE REVIEW

File: src/jquery.uls.languagefilter.js [hunk 1]
```diff
@@ -167,76 +167,69 @@
...
+					.fail( this.resultHandler.bind( this, query, results ) );
 			} else {
-				this.resultHandler( query );
+				this.resultHandler( query, results );
 			}
 		},
 
 		searchAPI: function ( query ) {
-			var languageFilter = this;
+			return $.get( this.options.searchAPI, { search: query } ).then( function ( result ) {
```
Comment (santhoshtr) [inline src/jquery.uls.languagefilter.js]: If the search api fails, it results this error:

```
jquery.uls.languagefilter.js:254 Uncaught TypeError: languageName.substring is not a function
    at LanguageFilter.autofill (jquery.uls.languagefilter.js:254)
    at LanguageFilter.resultHandler (jquery.uls.languagefilter.js:228)
```

languageName comes as rejected Promise

File: src/jquery.uls.languagefilter.js [hunk 2]
```diff
@@ -167,76 +167,69 @@
...
+					results = [];
 
-			$.get( languageFilter.options.searchAPI, {
-				search: query
-			}, function ( result ) {
 				$.each( result.languagesearch, function ( code, name ) {
-					if ( languageFilter.resultCount === 0 ) {
-						// Autofill the first result.
-						languageFilter.autofill( code, name );
+					if ( this.options.languages[ code ] ) {
```
Comment (santhoshtr) [inline src/jquery.uls.languagefilter.js]: this.options -> languagefilter.options

Comment (Nikerabbit) [reply src/jquery.uls.languagefilter.js]: Why? there is `bind( this )` below.

File: src/jquery.uls.languagefilter.js (hunk 1 above)
Comment (Nikerabbit) [reply src/jquery.uls.languagefilter.js]: This is only possible if `languageName` is truthy but not a string.

I guess on line 189, the third actual parameter might come from the failed promise.

**Comment by santhoshtr** [comment]
looks good now. Tested too. Since this is non-trivial change in the way search works, shall I suggest having some edits in README to reflect it? Unrelated, you may need to rebase to fix the CI issue.

**Pull request merged**